wgsl/index.bs

```diff
@@ -3762,16 +3762,53 @@ statement
 </pre>
 
 
-# Functions TODO # {#functions}
+# Functions # {#functions}
 
-A function declaration may only occur at [=module scope=].
-The function name is available for use after its declaration, until the end of the program.
+A <dfn noexport>function</dfn> performs computational work when invoked.
 
-If the return type of the function is not the void type, then the last statement
-in the function body must be a return statement.
+A function is invoked in one of the following ways:
+* By evaluating a function call expression. See [[#function-call-expr]].
+* By executing a function call statement. See [[#function-call-statement]].
+* An [=entry point=] function is invoked by the WebGPU implementation to perform
+    the work of a [=shader stage=] in a [=pipeline=]. See [[#entry-points]]
+
+There are two kinds of functions:
+* A [=built-in function=] is provided by the [SHORTNAME] implementation,
+    and is always available to a [SHORTNAME] program.
+    See [[#builtin-functions]].
+* A <dfn noexport>user-defined function</dfn> is declared in a [SHORTNAME] program.
+
+## Declaring a user-defined function ## {#function-declaration-sec}
+
+A <dfn noexport>function declaration</dfn> creates a user-defined function, by specifying:
+* An optional set of attributes.
+* The name of the function.
+* The formal parameter list: an ordered sequence of zero
+    or more [=formal parameter=] declarations,
+    separated by commas, and
+    surrounded by parentheses.
+* An optional, possibly decorated, <dfn noexport>return type</dfn>.
+* The <dfn noexport>function body</dfn>.
+
+A function declaration must only occur at [=module scope=].
+The function name is [=in scope=] from the start of the formal parameter list
+until the end of the program.
+
+The function name must be different from the name of every other declared object, including
```

Comment:
#1472 is intended to relax this rule I believe.

Reply:
I'll add that as an issue, noting the URL of the issue.